src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -278,33 +342,72 @@ private void UpdateDoubleCustomTags(double value, ReadOnlySpan<KeyValuePair<stri
         {
             try
             {
-                var index = this.FindMetricAggregatorsCustomTag(tags);
-                if (index < 0)
+                int index;
+                string[] tagKeys;
+                object[] tagValues;
+                do
                 {
-                    // TODO: Measurement dropped due to MemoryPoint cap hit.
-                    return;
+                    index = this.FindMetricAggregatorsCustomTag(tags, out tagKeys, out tagValues);
+                    if (index < 0)
+                    {
+                        // TODO: Measurement dropped due to MemoryPoint cap hit.
+                        return;
+                    }
                 }
-
-                this.metricPoints[index].Update(value);
+                while (!this.TryUpdateMetricPoint(index, value, tagKeys, tagValues));
             }
             catch (Exception)
             {
                 // TODO: Measurement dropped due to internal exception.
             }
         }
 
-        private int FindMetricAggregatorsDefault(ReadOnlySpan<KeyValuePair<string, object>> tags)
+        private bool TryUpdateMetricPoint(int index, long value, string[] tagKeys, object[] tagValues)
+        {
+            ref var metricPoint = ref this.metricPoints[index];
+
+            // If the MetricPointStatus = CandidateForRemoval, it is possible that another thread freed and repurposed the MetricPoint.
+            // We must check that the point being updated is the expected one.
+            if (index == 0 || (StringArrayComparer.Equals(metricPoint.Keys, tagKeys) && ObjectArrayComparer.Equals(metricPoint.Values, tagValues)))
```

Comment:
after thinking a bit more on this, this one also doesn't guarantee safety, as this check + actual update - are not atomic. For example, if this thread got context switched and never got the chance to run again until a long time, some other thread could have freed up this spacem causing this thread to write into some else's MP :(